ghost/core/core/server/services/members/SingleUseTokenProvider.js

```diff
@@ -46,16 +51,46 @@ class SingleUseTokenProvider {
             });
         }
 
+        if (model.get('used_count') >= 3) {
+            throw new ValidationError({
+                message: 'Token expired'
+            });
+        }
+
         const createdAtEpoch = model.get('created_at').getTime();
+        const firstUsedAtEpoch = model.get('first_used_at')?.getTime() ?? createdAtEpoch;
 
+        // Is this token already used?
+        if (model.get('used_count') > 0) {
+            const timeSinceFirstUsage = Date.now() - firstUsedAtEpoch;
+
+            if (timeSinceFirstUsage > this.validityPeriodAfterUsage) {
+                throw new ValidationError({
+                    message: 'Token expired'
+                });
+            }
+        }
         const tokenLifetimeMilliseconds = Date.now() - createdAtEpoch;
 
-        if (tokenLifetimeMilliseconds > this.validity) {
+        if (tokenLifetimeMilliseconds > this.validityPeriod) {
             throw new ValidationError({
                 message: 'Token expired'
             });
         }
 
+        if (!model.get('first_used_at')) {
+            await model.save({
```

Comment:
Wouldn't that fetch the model again?

Reply:
It would, yes - but I just wanted to check the reasoning, generally we stick to those crud-plugin methods in Ghost. Am happy with this as-is, just wanted to make sure we're aware of why it's there etc...